components/script/dom/gpucanvascontext.rs

```diff
@@ -0,0 +1,177 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::GPUCanvasContextBinding::{
+    GPUCanvasContextMethods, GPUSwapChainDescriptor,
+};
+use crate::dom::bindings::codegen::Bindings::GPUDeviceBinding::GPUDeviceBinding::GPUDeviceMethods;
+use crate::dom::bindings::codegen::Bindings::GPUObjectBaseBinding::GPUObjectDescriptorBase;
+use crate::dom::bindings::codegen::Bindings::GPUTextureBinding::{
+    GPUExtent3D, GPUExtent3DDict, GPUTextureDescriptor, GPUTextureDimension, GPUTextureFormat,
+};
+use crate::dom::bindings::reflector::{reflect_dom_object, DomObject, Reflector};
+use crate::dom::bindings::root::{Dom, DomRoot, LayoutDom};
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::gpuswapchain::GPUSwapChain;
+use crate::dom::htmlcanvaselement::HTMLCanvasElement;
+use crate::dom::webglrenderingcontext::LayoutCanvasRenderingContextHelpers;
+use dom_struct::dom_struct;
+use euclid::default::Size2D;
+use ipc_channel::ipc;
+use script_layout_interface::HTMLCanvasDataSource;
+use std::cell::Cell;
+use webgpu::{WebGPU, WebGPURequest};
+
+#[dom_struct]
+pub struct GPUCanvasContext {
+    reflector_: Reflector,
+    #[ignore_malloc_size_of = "channels are hard"]
+    channel: WebGPU,
+    canvas: Dom<HTMLCanvasElement>,
+    size: Cell<Size2D<u32>>,
+    swap_chain: DomRefCell<Option<Dom<GPUSwapChain>>>,
+    #[ignore_malloc_size_of = "Defined in webrender"]
+    webrender_image: Cell<Option<webrender_api::ImageKey>>,
+    #[ignore_malloc_size_of = "Defined in webrender"]
+    external_image_id: webrender_api::ExternalImageId,
+}
+
+impl GPUCanvasContext {
+    fn new_inherited(canvas: &HTMLCanvasElement, size: Size2D<u32>, channel: WebGPU) -> Self {
+        let (sender, receiver) = ipc::channel().unwrap();
+        let _ = channel.0.send(WebGPURequest::CreateContext(sender));
+        let external_image_id = receiver.recv().unwrap();
```

Comment:
I'm afraid that won't be possible. `WebrenderExternalImageRegistry` which creates the external_image_id for WebGL, Media and now WebGPU rests in the parent process, so creating id won't be possible in the content process.

Reply:
We can, however, do one thing. Since `GetWebGPUChan` is only used while creating a new Context, we can get the `external_id` in the same message itself. Even if, in future, `GetWebGPUChan` is used by some other GPU object we could send a `bool` corresponding to whether a new `external_id` is needed or not.
But this would raise one concern- we would be blocking `Constellation` until it acquires a lock on the registry. Although the work behind the lock is trivial, I'm still not sure if we can afford the block.

Reply:
Does `WebrenderExternalImageRegistry` have to be in the parent process?

Reply:
Keeping it in the content process would mean having a separate registry for each content process, i.e. duplicate ids will be generated. I am not sure how those would be handled by the webrender.